salt/state.py

```diff
@@ -3768,7 +3768,10 @@ def render_highstate(self, matches):
         for saltenv, states in six.iteritems(matches):
             for sls_match in states:
                 try:
-                    statefiles = fnmatch.filter(self.avail[saltenv], sls_match)
+                    if saltenv not in self.avail and '__env__' in self.avail:
+                        statefiles = fnmatch.filter(self.avail['__env__'], sls_match)
+                    else:
+                        statefiles = fnmatch.filter(self.avail[saltenv], sls_match)
```

Comment:
The old code used a try/except, but your changes introduce a membership check, which seems to make the try/except redundant. In cases where the saltenv is not present and `__env__` is not configured in `file_roots`, we do both a membership check _and_ a lookup, which we _know will fail_, and we also incur the overhead of raising and handling the exception. IMO it's better to use the results of the membership check to determine whether or not the saltenv could be found, perhaps like this:

```python
if saltenv in self.avail:
    statefiles = fnmatch.filter(self.avail[saltenv], sls_match)
else:
    if '__env__' in self.avail:
        statefiles = fnmatch.filter(self.avail['__env__'], sls_match)
    else:
        all_errors.append(
            'No matching salt environment for environment '
             '\'{0}\' found'.format(saltenv)
        )
```